modules/standard/DistributedBag.chpl

```diff
@@ -0,0 +1,1076 @@
+        ref segment = getPrivatizedThis.bag.segments[segmentIdx];
+        var nElems = segment.nElems.read() : int;
+        if nElems > average {
+          var take = nElems - average;
+          var tmpBuffer = buffer + bufferOffset;
+          segment.transferElements(tmpBuffer, take, buffer.locale.id);
+          bufferOffset += take;
+        }
+      }
+
+      // With the excess elements, redistribute it...
+      // BUG: Need to use Chapel Arrays due to issues with compiler automatically
+      // dereferencing a C_Ptr resulting in a segfault.
+      // BUG: Without '--no-denormalization' flag, the compiler will incorrectly
+      // use everything beyond this point by reference and *not* dereference it
+      // beforehand, causing more undefined behavior. For now, I use Chapel Arrays.
+      bufferOffset = 0;
+      for loc in localThis.targetLocales do on loc {
+        var average = avg;
+        ref segment = getPrivatizedThis.bag.segments[segmentIdx];
+        var nElems = segment.nElems.read() : int;
+        if average > nElems {
+          var give = average - nElems;
+          var arr : [1..give] eltType;
+          on bufferOffset {
+            var tmpBuffer = buffer;
+            for i in 1 .. give {
+              arr[i] = tmpBuffer[bufferOffset];
+              bufferOffset += 1;
+            }
+          }
+          for i in 1 .. give {
+            segment.addElements(arr[i]);
+          }
+        }
+      }
+
+      // Lastly, if there are items left over, just add them to our locale's segment.
+      if excess > bufferOffset {
+        ref segment = localThis.bag.segments[segmentIdx];
+        var give = excess - bufferOffset;
+        var tmpBuffer = buffer + bufferOffset;
+        segment.addElementsPtr(tmpBuffer, give, buffer.locale.id);
+      }
+
+      c_free(buffer);
+    }
+
+    // Phase 3: Release all locks from first node and segment to last node and segment.
+    for loc in localThis.targetLocales do on loc {
+      var instance = getPrivatizedThis;
+      for segmentIdx in 0 .. #here.maxTaskPar {
+        ref segment = instance.bag.segments[segmentIdx];
+        segment.releaseStatus();
+      }
+    }
+  }
+
+  /*
+    Iterate over each bag in each node. To avoid holding onto locks, we take
+    a snapshot approach, increasing memory consumption but also increasing parallelism.
+    This allows other concurrent, even mutating, operations while iterating,
+    but opens the possibility to iterating over duplicates or missing elements
+    from concurrent operations.
+
+    **Warning:** Breaking from an iterator will leak the snapshot due to issue #6912.
+    It will however leave all Bags in a safe state.
+  */
+  iter these() : eltType {
+    for loc in targetLocales {
+      for segmentIdx in 0..#here.maxTaskPar {
+        // The size of the snapshot is only known once we have the lock.
+        // BUG: c_ptr does not seem to work here and causes a segmentation fault at runtime.

```

Comment:
well yeah... c_ptr is always local and doesn't support use from another locale. _ddata would do that, but it's almost certainly better to just use a Chapel array in the first place.